broadcast/src/linked/signer/actor.rs

```diff
@@ -0,0 +1,1010 @@
+                }
+                std::cmp::Ordering::Equal => {
+                    // Ensure this matches the tip if the height is the same
+                    if chunk_tip.payload_digest != chunk.payload_digest {
+                        return Err(Error::ChunkMismatch(chunk.sequencer, chunk.height));
+                    }
+                }
+                std::cmp::Ordering::Greater => {}
+            }
+        }
+
+        // Verify digest
+        if D::read_from(&mut chunk.payload_digest.clone()).is_err() {
+            return Err(Error::InvalidDigest);
+        }
+
+        Ok(chunk)
+    }
+
+    ////////////////////////////////////////
+    // Journal
+    ////////////////////////////////////////
+
+    /// Returns the section of the journal for the given height.
+    fn get_journal_section(&self, height: u64) -> u64 {
+        height / self.journal_entries_per_section
+    }
+
+    /// Ensures the journal exists and is initialized for the given sequencer.
+    /// If the journal does not exist, it is created and replayed.
+    /// Else, no action is taken.
+    async fn journal_prepare(&mut self, sequencer: &PublicKey) {
+        // Return early if the journal already exists
+        if self.journals.contains_key(sequencer) {
+            return;
+        }
+
+        // Initialize journal
+        let cfg = journal::Config {
+            registry: Arc::new(Mutex::new(Registry::default())),
+            partition: (self.journal_naming_fn)(sequencer),
+        };
+        let mut journal = Journal::init(self.runtime.clone(), cfg)
+            .await
+            .expect("unable to init journal");
+
+        // Replay journal
+        {
+            debug!(sequencer = hex(sequencer), "Replaying journal");
+
+            // Prepare the stream
+            let stream = journal
+                .replay(self.journal_replay_concurrency, None)
+                .await
+                .expect("unable to replay journal");
+            pin_mut!(stream);
+
+            // Read from the stream, which may be in arbitrary order.
+            // Remember the highest link height
+            let mut tip: Option<wire::Link> = None;
+            let mut num_items = 0;
+            while let Some(msg) = stream.next().await {
+                num_items += 1;
+                let (_, _, _, msg) = msg.expect("unable to decode journal message");
+                let link = wire::Link::decode(msg).expect("journal message is unexpected format");
+                let height = link.chunk.as_ref().unwrap().height;
+                match tip {
+                    None => {
+                        tip = Some(link);
+                    }
+                    Some(ref t) => {
+                        if height > t.chunk.as_ref().unwrap().height {
+                            tip = Some(link);
+                        }
+                    }
+                }
+            }
+
+            // Set the tip
+            if let Some(link) = tip.take() {
```

Comment:
Why not just call `tip_man` for each link in this replay? It seems like you are duplicating some logic that already exists there.

Reply:
Ah, yeah I thought that at first but I have a check that tips don't try and move backwards; so I put in a comment that the `tip_manager` will panic if tips are put in out-of-order (possible due to the out-of-order nature of the journal replay)